### PR TITLE
Made Node.children getter result readonly

### PR DESCRIPTION
**Pull request by pandamicro**
Re: https://github.com/cocos/cocos-engine/issues/12519

### Changelog

* Made Node.children getter result readonly

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/core/scene-graph/base-node.ts [hunk 1]
```diff
@@ -137,9 +137,9 @@ export class BaseNode extends CCObject implements ISchedulable {
      * @readOnly
      */
     @editable
-    get children (): Node[] {
+    get children (): ReadonlyArray<Node[]> {
```
Comment (yanjifa) [inline cocos/core/scene-graph/base-node.ts:140]: ReadonlyArray<Node>

Comment (yanjifa) [inline cocos/core/scene-graph/base-node.ts]: ```ReadonlyArray<Node>```

Comment (pandamicro) [reply cocos/core/scene-graph/base-node.ts]: Fixed, that's why I should never use github to modify my code directly 🤣

**Comment by dumganhar** [comment]
```
cocos/2d/components/rich-text.ts#L766
Property 'splice' does not exist on type 'readonly Node[]'. Did you mean 'slice'?
```

**Comment by pandamicro** [comment]
@dumganhar It's a bug exposed by this PR right away, but I found it awkward that we can't actually reset the children to our Node, so still thinking what to do with it

**Comment by SantyWang** [comment]
RichText hack already removed.

**Review by SantyWang — APPROVED** [review]

**Comment by dumganhar** [comment]
> RichText hack already removed.

@pandamicro Try to sync the latest code from v3.6.1 branch

**Review by dumganhar — APPROVED** [review]

**Pull request merged**